### ISSUE TITLE
Xamarin project template

### ISSUE THREAD

**Issue by iongion** [bridgedotnet/Bridge#2367]
Pretty rough experience on OSX, can Xamarin can get project templates by using this:

http://www.monodevelop.com/developers/articles/project-templates

I am in my second day of trial and still stuck at a weird "compiler error 1" - most likely due to my bad set-up.

**Comment by geoffreymcgill** [comment]
Hi @iongion, 

Welcome to Bridge!

Thanks for bringing this to our attention. We did have this working in the past, but I don't think we have tested in **Xamarin Studio** nor **Mono Develop** for a while.

I believe _compiler error 1_ issues are generally related to pathing problems.

We'll run some experiments with on Mac and Linux with various IDE's.

**Comment by geoffreymcgill** [comment]
Were you trying to make a Xamarin Project Template, or just trying to compile the main Bridge Solution in Mono Develop or Xamarin Studio?

The main Bridge Solution is now compiling properly in Xamarin Studio (Mono 4.x) and available in the [Issue2367](https://github.com/bridgedotnet/Bridge/tree/Issue2367) branch. The NuGet packages are generated, and I was able to use those NuGet packages in a Visual Studio (Windows) project. Everything compiled perfectly. 

If you've started development of a Xamarin Project Template and willing to share that code, I could help get it working.

I also tested the Bridge Solution against the Visual Studio Mac Preview release (Mono 5), and there appears to be some substantial changes in Mono 5, when compared to Mono 4. I think we're going to wait for an RC release of VS Mac + Mono 5 before testing again.

**Comment by michaelcheers** [comment]
Related to #2646.

**Comment by cra0zy** [comment]
I'll get this done, just tell me in which folder/repo you want the addin project placed so I know when submitting the PR.

**Comment by cra0zy** [comment]
Addin done: https://nofile.io/f/UX6VxsaUr7N/MonoDevelop.Bridge.Addin_1.0.mpack

![bridgenettemplate](https://user-images.githubusercontent.com/6773302/30821009-0ca66560-a224-11e7-805d-3dba868584be.png)

Still have no idea which folder you use for templates (⊙.☉)7

**Comment by geoffreymcgill** [comment]
The VS template project is private, so we don't actually add any of these to the main Bridge project. That said, we could add this all into a new `templates` folder inside of [.build](https://github.com/bridgedotnet/Bridge/tree/master/.build).

@cra0zy – Do you know if this template will work with Visual Studio Mac too?

**Comment by cra0zy** [comment]
> @cra0zy – Do you know if this template will work with Visual Studio Mac too?

Actually the template is done for MonoDevelop 7, so yea, it will run on the newest version of VS for Mac.

The problem with MonoDevelop Addins is that the maintainers of MonoDevelop break compatibility with each major version released....

**Comment by geoffreymcgill** [comment]
We tested Bridge with the recent Visual Studio for Mac release and were able to add Bridge to the project using NuGet. More details:

https://github.com/bridgedotnet/Bridge/issues/2646#issuecomment-482894684

**Comment by geoffreymcgill** [comment]
With the current release, I don't see a way for custom **Project Templates** to be added to **Visual Studio for Mac**. Maybe they'll add this feature in the future. At least for now, I can't figure out a way to add custom **Project Templates**. 

We're also testing a new vsix package for the Visual Studio Gallery that supports Visual Studio 2019, see https://github.com/bridgedotnet/Bridge/issues/3910#issuecomment-482721414. 

Once we get some community feedback on the new installer we will publish to the Visual Studio Gallery and maybe that will get supported in VS Mac soon.

**Comment by cra0zy** [comment]
Unless they changed the template system, #3175 should have a perfectly usable custom Bridge template.

I can update the PR if you are interested in merging it, it really just needs an update to the nuget packages and an update to the target MonoDevelop version.

**Comment by geoffreymcgill** [comment]
@cra0zy – Give it a shot and I will test. The best I can tell, Visual Studio for Mac requires a special type Extension in order to add a custom Project Template. Doesn't look like VS Mac supports vsix extensions.

**Comment by cra0zy** [comment]
Yes, they have their own .mpack extensions and a custom template system: https://github.com/cra0zy/Bridge/blob/xamaddin/.build/templates/MonoDevelop.Bridge.Addin/Templates/Projects/BridgeClassLibrary.xpt.xml

Btw. Have you thought about creating `dotnet new` templates for Bridge? Both VS for Windows and Mac can use them as source templates in the extensions, and Rider can even directly load any system installed templates?

**Comment by cra0zy** [comment]
For reference on about question:
- https://github.com/dotnet/templating
- https://github.com/dotnet/templating/wiki/Available-templates-for-dotnet-new

**Comment by geoffreymcgill** [comment]
> Btw. Have you thought about creating dotnet new templates for Bridge? 

Yes, that's a feature we're going to try and get working soon.

**Comment by cra0zy** [comment]
I haven't forgotten about this, its just that MonoDevelop 8 is still not available for Linux and building from source would take quite some time...

**Comment by fabriciomurta** [comment]
It turned out making a `dotnet new` template for Bridge is not that simple. As far as creating the template, that's alright and straightforward. The problem lies where `dotnet build` does not work with Bridge. The dotnet CLI itself should work with .NET Framework projects but, the problem lies at least with the build tasks, as Bridge requires dependencies (assembly references) not exposed by the CLI (or compatible in its context).

What's possible right now is a `dotnet new` template that works, but then to build, either use `bridge build` (from [Bridge CLI](https://github.com/bridgedotnet/CLI) or Visual Studio. So switching tools would be required until the build process/references is figured out.

**Comment by cra0zy** [comment]
Just because `dotnet new` creates a template does not mean that `dotnet build` should build the project, they are completely separate tools. There are a bunch of templates that can get generated with `dotnet new` that don't support .NET Core.

PS. There exist 3 msbuild systems on linux, `dotnet msbuild` (linked only against .NET Core), `msbuild` and `xbuild`, in most cases you want to compile the project from command line by calling `msbuild` as its the one that has the most compatibility with everything.

**Comment by fabriciomurta** [comment]
Thanks for the feedback! Then it shouldn't be a problem at all to have the template up.
